File: tests/fixtures/discrimination/go/bottom_tier/terrible_14.go
package sample

import "fmt"

var registry = make(map[string]func(int) int)

func Register(name string, fn func(int) int) {
	registry[name] = fn
}

func Run(name string, x int) int {
	fn, _ := registry[name]
	if fn != nil {
		return fn(x)
	}
	return -1
}

func SetupRegistry() {
	Register("f1", func(x int) int { return x*2 + 3 })
	Register("f2", func(x int) int { return x*3 + 4 })
	Register("f3", func(x int) int { return x*4 + 5 })
	Register("f4", func(x int) int { return x*5 + 6 })
	Register("f5", func(x int) int { return x*6 + 7 })
	Register("f6", func(x int) int { return x*7 + 8 })
	Register("f7", func(x int) int { return x*8 + 9 })
	Register("f8", func(x int) int { return x*9 + 10 })
	Register("f9", func(x int) int { return x * x })
	Register("f10", func(x int) int { return x * x * x })
}

func RunAll(x int) string {
	SetupRegistry()
	t := 0
	t = t + Run("f1", x)
	t = t + Run("f2", x)
	t = t + Run("f3", x)
	t = t + Run("f4", x)
	t = t + Run("f5", x)
	t = t + Run("f6", x)
	t = t + Run("f7", x)
	t = t + Run("f8", x)
	t = t + Run("f9", x)
	t = t + Run("f10", x)
	return fmt.Sprintf("total=%d", t)
}

func RunSequence(names []string, x int) []int {
	SetupRegistry()
	results := make([]int, 0)
	for i := 0; i < len(names); i++ {
		results = append(results, Run(names[i], x))
	}
	return results
}

File: tests/fixtures/discrimination/python/bottom_tier/terrible_14.py
# eval-like dispatch, exec abuse, dynamic variable creation
import math

REGISTRY = {}


def register(name, code):
    REGISTRY[name] = code


def run(name, x):
    try:
        loc = {"x": x, "math": math, "result": 0}
        exec(REGISTRY[name], {}, loc)
        return loc["result"]
    except:
        return -1


def build_functions():
    register("f1", "result = x * 2 + 3")
    register("f2", "result = x * 3 + 4")
    register("f3", "result = x * 4 + 5")
    register("f4", "result = x * 5 + 6")
    register("f5", "result = x * 6 + 7")
    register("f6", "result = x * 7 + 8")
    register("f7", "result = x * 8 + 9")
    register("f8", "result = x * 9 + 10")
    register("f9", "result = x ** 2")
    register("f10", "result = x ** 3")


def run_all(x):
    build_functions()
    t = 0
    t = t + run("f1", x)
    t = t + run("f2", x)
    t = t + run("f3", x)
    t = t + run("f4", x)
    t = t + run("f5", x)
    t = t + run("f6", x)
    t = t + run("f7", x)
    t = t + run("f8", x)
    t = t + run("f9", x)
    t = t + run("f10", x)
    return t


def make_vars(n):
    d = {}
    for i in range(n):
        d["var_" + str(i)] = i * 47 + 13
    return d
